[AIP-127] gRPC Transcoding (#257)
googleapis/aip: aip/0127.md

---
aip:
  id: 127
  state: reviewing
  created: 2019-08-22
permalink: /127
redirect_from:
  - /0127
---

> # HTTP & gRPC Transcoding

Reviewer: HTTP and (let's avoid ampersands in titles)

Author: ```suggestion
# HTTP and gRPC Transcoding
```

APIs that follow [resource-oriented design][aip-121] are defined as [RPC][]
APIs, but the resource-oriented design framework allows those APIs to also be
presented as APIs that largely follow REST/JSON conventions. This is important

> in order to give developers cognitive leverage: over 80% of the public APIs

Reviewer 2: That's really "over 80% of the public Google APIs" - we've been trying to be less Google-centric, haven't we? I wonder whether we should just ditch this justification.

Author: No, it is 80% of public APIs worldwide, according to the slide from ProgrammableWeb that we use in the design class. (My gut instinct is that the 80% number is at least roughly correct.)

available follow most REST conventions, and developers are accustomed to that
pattern.

## Guidance

APIs **must** provide HTTP definitions for each RPC that they define, except

> for bi-directional streaming RPCs, which can not be natively supported using

Reviewer 2: "can not" => "cannot"?

Author: _Stet._

> HTTP/1.1. When providing a bi-directional streaming method, an API **must**

Reviewer: must or should? What if it's only useful as streaming?

Author: ```suggestion
HTTP/1.1. When providing a bi-directional streaming method, an API **should**
```

also offer an alternative method that does not rely on bi-directional
streaming.

### HTTP method and path

When using protocol buffers, each RPC **must** define the HTTP method and path
using the `google.api.http` annotation:

```proto
rpc CreateBook(CreateBookRequest) returns (Book) {
  option (google.api.http) = {
    post: "/v1/{parent=publishers/*}/books/*"
    body: "book"
  };
}

message CreateBookRequest {
  // The publisher who will publish this book.
  // When using HTTP/JSON, this field is automatically populated based
  // on the URI, because of the `{parent=publishers/*}` syntax.
  string parent = 1;

  // The book to create.
  // When using HTTP/JSON, this field is populated based on the HTTP body,
  // because of the `body: "book"` syntax.
  Book book = 2;

  // The user-specified ID for the book.
  // When using HTTP/JSON, this field is populated based on a query string
  // argument, such as `?book_id=foo`. This is the fallback for fields that
  // are not included in either the URI or the body.
  string book_id = 3;
}
```

- The first key (`post` in this example) corresponds to the HTTP method. RPCs

>   **may** use `get`, `post`, `patch`, and `delete`.

Reviewer: s/and/or/

Author: ```suggestion
  **may** use `get`, `post`, `patch`, or `delete`.
```

  - RPCs **must** use the prescribed HTTP verb for each standard method, as
    discussed in [AIP-131][], [AIP-132][], [AIP-133][], [AIP-134][], and
    [AIP-135][]
  - RPCs **should** use the prescribed HTTP verb for custom methods, as
    discussed in [AIP-136][].
  - RPCs **should not** use `put` or `custom`.
- The corresponding value represents the URI.
  - URIs **must** use the `{foo=bar/*}` syntax to represent a variable that
    should be populated in the request proto. When extracting a [resource
    name][aip-122], the variable **must** include the entire resource name, not
    just the ID component.
  - URIs **must** use the `*` character to represent ID components, which
    matches all URI-safe characters except for `/`. URIs **may** use `**` as

>     the final segment of a URI if matching `/` is required.

Reviewer: I believe we support ** elsewhere now also...? We should check with the team on this.

Author: Even if we do, I assert it is almost always a bad idea.

> - The `body` key defines which field in the request is sent as the HTTP body.

Reviewer: s/is sent/will be sent/ ?

Author: ```suggestion
- The `body` key defines which field in the request will be sent sent as the HTTP body.
```

  If the body is `*`, then this indicates that the request object itself is the
  HTTP body.

>   - RPCs **must not** define a `body` at all for GET or DELETE requests.

Reviewer: s/GET/Get [AIP-???]/, s/DELETE/Delete [AIP-???]/ ? (to be consistent with the following line.

Author: This is about the HTTP verb, not the standard methods. A custom method that uses the GET verb also must not have a `body`. I will reword this to make that clear.

Author: Reworded.

  - RPCs **must** use the prescribed `body` for Create ([AIP-133][]) and Update
    ([AIP-134][]) requests.
  - RPCs **should** use the prescribed `body` for custom methods ([AIP-136][]).

### Multiple URI bindings

> Occasionally, an RPC needs to correspond to more than one HTTP path:

Reviewer: s/HTTP path/URI/ ?

Author: ```suggestion
Occasionally, an RPC needs to correspond to more than one URI:
```

```proto
rpc CreateBook(CreateBookRequest) returns (Book) {
  option (google.api.http) = {
    post: "/v1/{parent=publishers/*}/books"
    body: "book"
    additional_bindings: {
      post: "/v1/{parent=authors/*}/books"
      body: "book"
    }

>     additional_bindings: {

Reviewer: Maybe chop this one out? This opens the door to questions about Creating top-level resources (no parent)

Author: Discussed on GVC; this one is intentional.

      post: "/v1/books"
      body: "book"
    }
  }
}
```

- RPCs **may** define any number of additional bindings. The structure is
  identical to the `google.api.http` annotation (in fact, it is a recursive
  reference).
- RPCs **must not** define an additional binding within an additional binding.
- The `body` clause **must** be identical in the top-level annotation and each

>   additional binding.

Reviewer 2: It feels like a bit of a hole in tooling/implementation that the body even needs to be specified multiple times if it's always got to be the same. Would it *ever* not be the same, even in exceptional circumstances?

Author: I agree, but that is where the tooling is right now. (I was thinking about that when I wrote this.)
